### PR TITLE
JavaScript Variable Scope and Hoisting

### PR DESCRIPTION
**Pull request by AdamChrimes**
## Description

This is a potential fix for https://github.com/nhsuk/nhsuk-frontend/issues/398

Globally scoped variables were causing issues with JavaScript frameworks such as Angular, so they have been moved to be locally scoped within functions. However, this could be improved as we have to duplicate variables between the multiple functions.

There was also an issue with the autocomplete load event listener with Angular, The wrap-search element was not being found as the page load event listener was coming after the search for the element which caused issues in JavaScript frameworks such as Angular.

Currently testing this out with Angular and ES5/Webpack, also looking for better ways to do this and potentially look for advice on how we write our JavaScript (ES6).

### Further reading

https://www.sitepoint.com/demystifying-javascript-variable-scope-hoisting/

### To Do:

- [x] Tested in Angular using the [NhsukFrontendAngular](https://github.com/puravupadhyay/nhsuk-frontend-angular) repo provided by @puravupadhyay
- [x] Tested in Browsers (to IE8) with ES5 (using the nhsuk-frontend repo pages)
- [ ] Tested in third party application (non JS framework) which uses the `nhsuk-frontend` library as a npm package
- [ ] Tested in third party application (non JS framework) which uses the `nhsuk-frontend` library from compiled files
- [ ] Investigate better ways to write this JavaScript to avoid duplication of variables, also is there a better way to check the element exists eg `if (element)` or an alternative not to have to do this?

The above also needs to be tested by someone else to confirm.

### CODE REVIEW

File: packages/components/header/header.js [hunk 1]
```diff
@@ -21,25 +21,29 @@ function toggleClass(ele, class1) {
...
-const searchContainer = document.querySelector('#wrap-search');
-const menuSearchContainer = document.querySelector('#content-header');
-
 function toggleSearch(e) {
+  const searchToggleButton = document.querySelector('#toggle-search');
+  const searchContainer = document.querySelector('#wrap-search');
+  const menuSearchContainer = document.querySelector('#content-header');
+
   e.preventDefault();
+  e.stopImmediatePropagation();
```
Comment (mikemonteith) [inline packages/components/header/header.js]: Need to take these out (1)

File: packages/components/header/header.js [hunk 2]
```diff
@@ -50,22 +54,27 @@ function handleSearchToggle() {
...
-const menuToggleButton = document.querySelector('#toggle-menu');
-const menuClose = document.querySelector('#close-menu');
-const nav = document.querySelector('#header-navigation');
-
 function toggleMenu(e) {
+  const menuToggleButton = document.querySelector('#toggle-menu');
+  const nav = document.querySelector('#header-navigation');
+
   e.preventDefault();
+  e.stopImmediatePropagation();
```
Comment (mikemonteith) [inline packages/components/header/header.js]: (2)

**Comment by AdamChrimes** [comment]
@mikemonteith removed the stopImmediatePropogation calls.

File: package-lock.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "nhsuk-frontend",
-  "version": "1.0.0",
+  "version": "1.1.0",
```
Comment (mikemonteith) [inline package-lock.json]: Does this need to be a minor version increase? It's a bugfix rather than new feature so I'd make it a patch update.

**Pull request merged**